client/src/procedures/primitives.rs

```diff
@@ -789,3 +798,115 @@ impl UseSecret for AeadDecrypt {
+    /// The location of the shared secret `z`.
+    pub shared_secret: Location,
+    /// The number of bytes of key material that should be derived.
+    pub key_len: usize,
+    /// Agreement PartyUInfo
+    pub apu: Vec<u8>,
+    /// Agreement PartyVInfo
+    pub apv: Vec<u8>,
+    /// The location to write the derived key material into.
+    pub output: Location,
+}
+
+impl DeriveSecret for ConcatKdf {
+    type Output = ();
+
+    fn derive(self, guard: Buffer<u8>) -> Result<Products<()>, FatalProcedureError> {
+        let derived_key_material: Vec<u8> = match self.hash {
+            Sha2Hash::Sha256 => Self::concat_kdf(
+                Sha256::new(),
+                self.algorithm_id.as_ref(),
+                self.key_len,
+                guard.borrow().as_ref(),
+                self.apu.as_ref(),
+                self.apv.as_ref(),
+            ),
+            Sha2Hash::Sha384 => Self::concat_kdf(
+                Sha384::new(),
+                self.algorithm_id.as_ref(),
+                self.key_len,
+                guard.borrow().as_ref(),
+                self.apu.as_ref(),
+                self.apv.as_ref(),
+            ),
+            Sha2Hash::Sha512 => Self::concat_kdf(
+                Sha512::new(),
+                self.algorithm_id.as_ref(),
+                self.key_len,
+                guard.borrow().as_ref(),
+                self.apu.as_ref(),
+                self.apv.as_ref(),
+            ),
+        }?;
+
+        Ok(Products {
+            secret: derived_key_material,
+            output: (),
+        })
+    }
+
+    fn source(&self) -> &Location {
+        &self.shared_secret
+    }
+
+    fn target(&self) -> &Location {
+        &self.output
+    }
+}
+
+impl ConcatKdf {
+    /// The Concat KDF as defined in Section 5.8.1 of NIST.800-56A.
+    pub fn concat_kdf<D: Digest>(
+        mut digest: D,
+        alg: &str,
+        len: usize,
+        z: &[u8],
+        apu: &[u8],
+        apv: &[u8],
+    ) -> Result<Vec<u8>, FatalProcedureError> {
+        let mut output: Vec<u8> = Vec::new();
+
+        let target: usize = (len + (D::output_size() - 1)) / D::output_size();
+        let rounds: u32 =
+            u32::try_from(target).map_err(|_| FatalProcedureError::from("u32 iteration overflow".to_owned()))?;
+
+        for count in 0..rounds {
+            // Iteration Count
+            digest.update(&(count as u32 + 1).to_be_bytes());
+
+            // Derived Secret
+            digest.update(z);
+
+            // AlgorithmId
+            digest.update(&(alg.len() as u32).to_be_bytes());
+            digest.update(alg.as_bytes());
+
+            // PartyUInfo
+            digest.update(&(apu.len() as u32).to_be_bytes());
+            digest.update(apu);
+
+            // PartyVInfo
+            digest.update(&(apv.len() as u32).to_be_bytes());
+            digest.update(apv);
+
+            // Shared Key Length
+            digest.update(&((len * 8) as u32).to_be_bytes());
+
```

Comment:
I can't find a reference to including the output length as part of the hash in the NIST PDF (it's not part of `OtherInfo` as far as I can see), nor any existing implementations such as Hyperledger Aries: https://github.com/hyperledger/aries-askar/blob/main/askar-crypto/src/kdf/concat.rs

Reply:
The `keydatalen` is included in the optional `SuppPubInfo` parameter for JWA with ECDH-ES specifically: https://www.rfc-editor.org/rfc/rfc7518.html#section-4.6.2

>    SuppPubInfo
>       This is set to the keydatalen represented as a 32-bit big-endian
>       integer.
> 
>    SuppPrivInfo
>       This is set to the empty octet sequence.

So it should not be hardcoded, and the procedure should be updated to allow the `SuppPubInfo` and `SuppPrivInfo` optional parameters.